src/Analyzers/CSharp/Analyzers/UseAutoProperty/CSharpUseAutoPropertyAnalyzer.cs

```diff
@@ -28,83 +39,65 @@ protected override bool SupportsPropertyInitializer(Compilation compilation)
             {
-                AnalyzeMembers(context, namespaceDeclaration.Members, analysisResults);
+                // An argument will disqualify a field if that field is used in a ref/out position.  
+                // We can't change such field references to be property references in C#.
+                if (argument.RefKindKeyword.Kind() != SyntaxKind.None)
+                    AddIneligibleFieldsForExpression(argument.Expression);
             }
-            else if (member is TypeDeclarationSyntax(
-                SyntaxKind.ClassDeclaration or
-                SyntaxKind.StructDeclaration or
-                SyntaxKind.RecordDeclaration or
-                SyntaxKind.RecordStructDeclaration) typeDeclaration)
+
+            foreach (var refExpression in codeBlock.DescendantNodesAndSelf().OfType<RefExpressionSyntax>())
+                AddIneligibleFieldsForExpression(refExpression.Expression);
+
+            // Can't take the address of an auto-prop.  So disallow for fields that we do `&x` on.
+            foreach (var addressOfExpression in codeBlock.DescendantNodesAndSelf().OfType<PrefixUnaryExpressionSyntax>())
             {
-                // If we have a class or struct, recurse inwards.
-                AnalyzeMembers(context, typeDeclaration.Members, analysisResults);
+                if (addressOfExpression.Kind() == SyntaxKind.AddressOfExpression)
+                    AddIneligibleFieldsForExpression(addressOfExpression.Operand);
             }
-            else if (member is PropertyDeclarationSyntax propertyDeclaration)
+
+            foreach (var memberAccess in codeBlock.DescendantNodesAndSelf().OfType<MemberAccessExpressionSyntax>())
             {
-                AnalyzeProperty(context, propertyDeclaration, analysisResults);
+                if (CouldReferenceField(memberAccess))
+                    AddIneligibleFieldsIfAccessedOffNotDefinitelyAssignedValue(semanticModel, memberAccess, ineligibleFields, cancellationToken);
             }
-        }
 
-        protected override void RegisterIneligibleFieldsAction(
-            List<AnalysisResult> analysisResults, HashSet<IFieldSymbol> ineligibleFields,
-            Compilation compilation, CancellationToken cancellationToken)
-        {
-            var groups = analysisResults.Select(r => (typeDeclaration: (TypeDeclarationSyntax)r.PropertyDeclaration.Parent!, r.SemanticModel))
-                                        .Distinct()
-                                        .GroupBy(n => n.typeDeclaration.SyntaxTree);
+            bool CouldReferenceField(ExpressionSyntax expression)
+            {
+                // Don't bother binding if the expression isn't even referencing the name of a field we know about.
+                var rightmostName = expression.GetRightmostName()?.Identifier.ValueText;
+                return rightmostName != null && fieldNames.Contains(rightmostName);
+            }
```

Comment:
we collect the names of fields in the type at the start, greatly reducing the amount of binding we actually have to do in the type.  Now we only bind expressions actually using that name, and nothing else.